stdlib/public/core/Filter.swift.gyb

```diff
@@ -295,6 +326,30 @@ public struct ${Self}<
     } while index != _base.endIndex && !_predicate(_base[index])
   }
 
+%   if Traversal == 'Bidirectional':
+  /// Returns the previous `index` of an element that `self._predicate` matches
+  /// otherwise `self._base.startIndex`
+  @inline(__always)
+  internal func _previousFiltered(_ index: Base.Index) -> Base.Index {
+    var index = index
+    _previousFilteredInPlace(&index)
+    return index
+  }
+
+  /// Returns the first index `i` preceding `index` where either
+  /// `_predicate(base[i]) == true` or `i == _base.startIndex`.
+  ///
+  /// - Precondition: `index != startIndex`
+  /// - Postcondition: `i < index`
+  @inline(__always)
+  internal func _previousFilteredInPlace(_ index: inout Base.Index) {
+    _precondition(index != _base.startIndex, "can't retreat before startIndex")
+    repeat {
+      _base.formIndex(before: &index)
+    } while index != _base.startIndex && !_predicate(_base[index])
```

Comment:
And the underlying collection should be doing range checks anyway.  The precondition in the forward movement is just a QoI thing, not required for memory safety.